pygame/joystick.py

```diff
@@ -0,0 +1,197 @@
+
+from pygame._error import SDLError
+from pygame._sdl import ffi, sdl
+from pygame.base import register_quit
+
+
+def _check_init():
+    if not sdl.SDL_WasInit(sdl.SDL_INIT_JOYSTICK):
+        raise SDLError("joystick system not initialized")
+
+
+def autoinit():
+    if not sdl.SDL_WasInit(sdl.SDL_INIT_JOYSTICK):
+        if sdl.SDL_InitSubSystem(sdl.SDL_INIT_JOYSTICK):
+            return False
+        sdl.SDL_JoystickEventState(sdl.SDL_ENABLE)
+        register_quit(autoquit)
+    return True
+
+
+def autoquit():
+    for joydata in Joystick._OPEN_JOYSTICKS:
+        sdl.SDL_JoystickClose(joydata)
+    del Joystick._JOYSTICKS[:]
+
+    if sdl.SDL_WasInit(sdl.SDL_INIT_JOYSTICK):
+        sdl.SDL_JoystickEventState(sdl.SDL_DISABLE)
+        sdl.SDL_QuitSubSystem(sdl.SDL_INIT_JOYSTICK)
+
+
+def init():
+    """ init() -> None
+    Initialize the joystick module.
+    """
+    if not autoinit():
+        SDLError.from_sdl_error()
+
+
+def quit():
+    """ quit() -> None
+    Uninitialize the joystick module.
+    """
+    autoquit()
+
+
+def get_init():
+    """ get_init() -> bool
+    Returns True if the joystick module is initialized.
+    """
+    return sdl.SDL_WasInit(sdl.SDL_INIT_JOYSTICK) != 0
+
+
+def get_count():
+    """ get_count() -> count
+    Returns the number of joysticks.
+    """
+    _check_init()
+    return sdl.SDL_NumJoysticks()
+
+
+class Joystick(object):
+    """ Joystick(id) -> Joystick
+    Create a new Joystick object.
+    """
+
+    _OPEN_JOYSTICKS = {}
+
+    def __init__(self, i):
+        _check_init()
+        if i < 0 or i >= get_count():
+            raise SDLError("Invalid joystick device number")
+        self._id = i
+
+    @property
+    def _joydata(self):
+        joydata = self._OPEN_JOYSTICKS.get(self._id)
+        if not joydata:
+            raise SDLError("Joystick not initialized")
+        return joydata
+
+    def init(self):
+        """ init() -> None
+        initialize the Joystick
+        """
+        if self._id not in self._OPEN_JOYSTICKS:
+            joydata = sdl.SDL_JoystickOpen(self._id)
+            if not joydata:
+                raise SDLError.from_sdl_error()
+            self._OPEN_JOYSTICKS[self._id] = joydata
+
+    def quit(self):
+        """ quit() -> None
+        uninitialize the Joystick
+        """
+        joydata = self._OPEN_JOYSTICKS.get(self._id)
+        if joydata:
+            sdl.SDL_JoystickClose(joydata)
+            del self._OPEN_JOYSTICKS[self._id]
+
+    def get_init(self):
+        """ get_init() -> bool
+        check if the Joystick is initialized
+        """
+        return self._OPEN_JOYSTICKS.get(self._id) is not None
+
+    def get_id(self):
+        """ get_id() -> int
+        get the Joystick ID
+        """
+        return self._id
+
+    def get_name(self):
+        """ get_name() -> string
+        get the Joystick system name
+        """
+        return sdl.SDL_JoystickName(self._id)
```

Comment:
This should be wrapped in a ffi.string(), otherwise the result really isn't that useful